STM32WBAX PM rework to support STOP/STDBY modes

asm5878 opened the pull request:
Aim of this PR is a cleaner support of stop/stdby modes for stm32wbax socs.
System power management has been reworked to remove:
-  the dependency on scm (system clock manager) module.
- all the assumptions on the BLE hci driver activation

Scope of this PR has been reduced to manage only the pm at system level.
A new PR (https://github.com/zephyrproject-rtos/zephyr/pull/95151) has been created to update the pm for the stm32wbax ble hci driver.
 
FYI: @Chastillon

asm5878 committed:
soc: st: stm32: stm32wbax pm enhancements

soc power management rework to support power states
removing dependency on ST system clock manager (SCM).

Signed-off-by: Alessandro Manganaro <alessandro.manganaro@st.com>

msmttchr reviewed (commented):
Hello @asm5878, thanks a lot for this draft PR, very interesting proposal. I added few preliminary comments, and I would like also to involve @HoZHel and @FrancescoCiminoST

asm5878 commented on samples/bluetooth/peripheral_hr/boards/nucleo_wba65ri.overlay:
```diff
@@ -0,0 +1,28 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2023 STMicroelectronics
+ */
+
+/ {
+	/* Change min residency time to ease power consumption measurement */
+	cpus {
+		cpu0: cpu@0 {
+			cpu-power-states = <&stop0 &stop1 &standby>;
+		};
+
+		power-states {
+			standby: state2 {
+				compatible = "zephyr,power-state";
+				power-state-name = "suspend-to-ram";
+				substate-id = <1>;
+				min-residency-us = <10000>;
+				exit-latency-us = <1000>;
+			};
+		};
+	};
+};
+
+&lptim1 {
+	status = "okay";
+};
```
Same comment as before

asm5878 replied:
same comment as before.

asm5878 commented on samples/bluetooth/peripheral_hr/prj.conf:
```diff
@@ -8,3 +8,9 @@ CONFIG_BT_BAS=y
 CONFIG_BT_HRS=y
 CONFIG_BT_DEVICE_NAME="Zephyr Heartrate Sensor"
 CONFIG_BT_DEVICE_APPEARANCE=833
+
+CONFIG_PM=y
+CONFIG_PM_DEVICE=y
+CONFIG_PM_DEVICE_RUNTIME=y
+CONFIG_PM_DEVICE_SYSTEM_MANAGED=y
+CONFIG_PM_S2RAM=y
```
We could just leave `CONFIG_PM` to be selected by user in each sample and use it to conditionally enable all the rest of configs in [nucleo_wba65ri_defconfig](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/st/nucleo_wba65ri/nucleo_wba65ri_defconfig)

asm5878 replied:
I didn't want to enable the suspend to ram option by default. So I have left the user the possibility to choice the STOP or STDBY. 
CONFIG_PM and CONFIG_PM_S2RAM can be selected in impendent way.  The other configs are implicitly activated.

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -19,7 +23,8 @@
 
 #ifdef CONFIG_BT_STM32WBA
-#include "scm.h"
+#include "linklayer_plat.h"
+#include "ll_sys.h"
```
Agreed, for sure an error during the rebase.

asm5878 replied:
Fixed

asm5878 commented on subsys/pm/pm.c:
```diff
@@ -19,6 +19,9 @@
 
 #include "pm_stats.h"
 #include "device_system_managed.h"
+#ifdef CONFIG_BT_STM32WBA
+#include "ll_sys.h"
+#endif
```
yes, this file must stay untouched compared to upstream

asm5878 replied:
No more modifications on pm.c

asm5878 commented on west.yml:
```diff
@@ -245,7 +245,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 126cbbee19208b7aaca5ad8b287cf104e8bc837a
+      revision: pull/294/head
```
Indeed when [PR 294](https://github.com/zephyrproject-rtos/hal_stm32/pull/294) will be approved we will have a new commit to be used

msmttchr commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -28,59 +33,71 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
 
-	/* Select mode entry : WFE or WFI and enter the CPU selected mode */
-	k_cpu_idle();
+	/* MCU enters in low-power modes */
+	basepri = __get_BASEPRI();
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
```
I think the reasoning here is the following:
1) When executing these power management procedures, we do not want any interrupt, regardless of his priority to preempt us
2) We would like any possible pending interrupt to transform the Wfi in a nop and skip the power mode entry 

For 1), we globally disable interrupts with `__disable_irq`
For 2), we unmask all interrupt priorities

The original code of Zephyr using `irq_lock`, allows effectively to preempt this critical function by high priority irq (priority lower  than `_EXC_IRQ_DEFAULT_PRIO`.

In turn, since interrupt are calling the pm management routines both in direct or regular case, in the middle of this critical routine, not only the high priority interrupt is executed, but also the `pm_system_resume` is executed effectively re-enabling all the interrupts.
See https://github.com/zephyrproject-rtos/zephyr/blob/553fc84e77e245eda41dd1e2f2ba6ed1055d4bcb/arch/arm/core/cortex_m/isr_wrapper.c#L39-L66

As such, we took inspiration from https://github.com/zephyrproject-rtos/zephyr/blob/553fc84e77e245eda41dd1e2f2ba6ed1055d4bcb/soc/nordic/nrf54h/power.c#L183-L208 for our implementation.

This implementation will remove the feature that allows high priority irq to preempt cpu in these critical routines, but I believe this feature would not work during s2ram procedure execution where being interrupted is a system crash guaranteed.

mathieuchopstm replied:
Thanks for the detailed explanation. As you say yourself, `IRQ_ZERO_LATENCY` is not used on STM32 platforms. Since it is (unless I'm misunderstanding things) the only way to have interrupts with priority higher than `_EXC_IRQ_DEFAULT_PRIO`, we should not be affected.

This does raise the point that despite the `arch_irq_lock()` specification being:
https://github.com/zephyrproject-rtos/zephyr/blob/1ee39b982396d2c42d77b537ca2ae46721bcbc3a/include/zephyr/irq.h#L223-L226
(Emphasis mine: _This routine disables **all** interrupts on the CPU_)

on ARM platforms, this is not respected: ZLI might occur under `irq_lock()`... but this is documented, and I would argue it's fine because ZLI should *not* interact with the OS?

Regardless, adding a `__disable_irq()` in `pm_state_set()` is at best a band-aid: there remains a gigantic window in `pm_system_suspend()` (between `arch_irq_lock()` in `idle.c` and `pm_state_set()` in SoC code) where an interrupt could occur. The fix, if any, would be introducing a new API set to distinguish between `disable ALL interrupts` and `disable <specified subset> interrupts` (e.g., `<specified subset> = all non-ZLI`)

-----

WRT my other remark, I still think `__ISB()` is not required.
ARMv8-M Architecture Reference Manual says:
```
Between any change to special purpose registers which require synchronization or memory-mapped registers and
a subsequent Context synchronization event, it is IMPLEMENTATION DEFINED whether an indirect read of the register by the PE uses the old or new values, and may vary with each use of the unsynchronized value.
```
and
```
Except for writes to the CONTROL register, any change to a special-purpose register by a CPS or MSR instruction
is guaranteed:
• Not to affect that CPS or MSR instruction, or any instruction preceding it in program order.
• To be visible to all instructions that appear in program order after the CPS or MSR.
```
(this latter paragraph is found verbatim in the ARMv7-M Architecture Reference Manual)

My understanding of these paragraphs is that writing to `BASEPRI` should not require any synchronization (`ISB`), regardless of what the comment in `idle.c` may say.

msmttchr replied:
Ok for me. @mathieuchopstm, is it OK for you as well ?

mathieuchopstm replied:
Fine by me.

msmttchr commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +114,150 @@ static void set_mode_suspend_to_ram(void)
 	LL_PWR_SetPowerMode(LL_PWR_MODE_STANDBY);
 
+	/* Save FPU state and configuration */
+	z_arm_save_fp_context(&fpu_state);
```
Good suggestion, I think you’re right.

msmttchr replied:
Now it should be OK, see #93057

msmttchr commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -28,59 +33,71 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
-		LOG_DBG("Unsupported power state substate-id %u", substate_id);
-		break;
-	}
+static void scb_suspend(SCB_Type *backup)
+{
+	backup->ICSR = SCB->ICSR;
+	backup->VTOR = SCB->VTOR;
+	backup->AIRCR = SCB->AIRCR;
+	backup->SCR = SCB->SCR;
+	backup->CCR = SCB->CCR;
+	memcpy((uint32_t *) backup->SHPR, (uint32_t *)SCB->SHPR, sizeof(SCB->SHPR));
+	backup->SHCSR = SCB->SHCSR;
+	backup->CFSR = SCB->CFSR;
+	backup->HFSR = SCB->HFSR;
+	backup->DFSR = SCB->DFSR;
+	backup->MMFAR = SCB->MMFAR;
+	backup->BFAR = SCB->BFAR;
+	backup->AFSR = SCB->AFSR;
+	backup->CPACR = SCB->CPACR;
+}
+static void scb_resume(SCB_Type *backup)
+{
+	SCB->ICSR = backup->ICSR;
+	SCB->VTOR = backup->VTOR;
+	SCB->AIRCR = backup->AIRCR;
+	SCB->SCR = backup->SCR;
+	SCB->CCR = backup->CCR;
+	memcpy((uint32_t *)SCB->SHPR, (uint32_t *)backup->SHPR, sizeof(SCB->SHPR));
+	SCB->SHCSR = backup->SHCSR;
+	SCB->CFSR = backup->CFSR;
+	SCB->HFSR = backup->HFSR;
+	SCB->DFSR = backup->DFSR;
+	SCB->MMFAR = backup->MMFAR;
+	SCB->BFAR = backup->BFAR;
+	SCB->AFSR = backup->AFSR;
+	SCB->CPACR = backup->CPACR;
```
Excellent comment, in my opinion. Valid also for fpu?
SCB is implementation specific, at least in some fields, like SHPR.
We used the CMSIS typedef `SCB_Type`.
I am not sure whether this can be moved at arch level, but I would be 100%, in favour, if it’s possible

msmttchr replied:
This code should not be neeeded, since https://github.com/zephyrproject-rtos/zephyr/pull/93699 introduced functions to save and restore SCB for CortexM architecture.

asm5878 replied:
Thanks @msmttchr. Fixed.

msmttchr commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +114,150 @@ static void set_mode_suspend_to_ram(void)
+	 */
+	z_arm_configure_static_mpu_regions();
+#endif /* CONFIG_ARM_MPU */
+	z_arm_fault_init();
```
```suggestion
```
This should not be needed since `SCB.CCR` is restored by `scb_resume`.

MBrandt-Z commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +110,133 @@ static void set_mode_suspend_to_ram(void)
 	} else {
-		scm_setwaitstates(RUN);
+		/* Apply waitsates for HSE32 configuration */
+		__HAL_FLASH_SET_LATENCY(FLASH_LATENCY_0);
```
Setting latency to 0 will lead to a hard fault, when running on higher clock rates (e.g. PLL at 100MHz)
Latency should only be changed to 1 while entering LP Modes if it is zero before, if not keep it as it is...

MBrandt-Z commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +110,133 @@ static void set_mode_suspend_to_ram(void)
+		stm32_clock_control_init(NULL);
+	} else {
+		/* Apply waitsates for HSE32 configuration */
+		__HAL_FLASH_SET_LATENCY(FLASH_LATENCY_0);
```
can we get here when leaving standby mode? If yes, setting latency to 0 may be the wrong value for th current clock. (see comment for set_mode_stop_exit() below)

asm5878 commented on subsys/pm/pm.c:
```diff
@@ -163,6 +169,17 @@ bool pm_system_suspend(int32_t kernel_ticks)
+	{
+		uint32_t radio_remaining_time, cmd_status;
+		int32_t radio_ticks;
+		cmd_status =
```
No more modifications on pm.c

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -28,59 +33,71 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+	/* MCU exits from low power mode and set back the basepri */
+	__set_BASEPRI(basepri);
 
 	return 0;
```
Fixed

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +114,150 @@ static void set_mode_suspend_to_ram(void)
-		LOG_DBG("Suspend to RAM needs CONFIG_PM_S2RAM to be enabled");
-#endif
+	case PM_STATE_SUSPEND_TO_RAM:
+		__disable_irq();
```
Could we assume this point has been already clarified and we don't need extra changes?

mathieuchopstm replied:
OK for me. I think there is a little change to be done for clarity but I'll post that as separate review comment.

erwango commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -63,7 +63,8 @@
 				substate-id = <1>;
 				min-residency-us = <1000>;
-				exit-latency-us = <50>;
+				/* From DS, Twu_stdby with retention: 51.49 */
+				exit-latency-us = <300>;
```
Link between comment and change is not crystal clear.

asm5878 replied:
done and this PR is now only related to system pm

erwango commented on samples/bluetooth/peripheral_hr/sample.yaml:
```diff
@@ -110,3 +110,17 @@ tests:
+      - nucleo_wba65ri
+      - nucleo_wba55cg
+    integration_platforms:
+      - nucleo_wba65ri
+      - nucleo_wba55cg
```
Aim of integration platform is to quickly validate on minimum number of boards. Pick up one.

asm5878 replied:
Fixed

erwango commented on samples/bluetooth/peripheral_hr/boards/stm32wba_stdby.overlay:
```diff
@@ -0,0 +1,23 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2023 STMicroelectronics
```
s/2023/2025

erwango reviewed (changes requested):
First commit should be split with finer granularity to allow proper review.
Maybe split in 2 PR with first PR enabling standby and second adding PM with radio activity ?

msmttchr reviewed (changes requested):
Some changes to reuse common SCB save/restore code.

etienne-lms commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +112,134 @@ static void set_mode_suspend_to_ram(void)
+		MODIFY_REG(RAMCFG_SRAM2->CR, RAMCFG_CR_WSC, RAMCFG_WAITSTATE_1);
+	}
+	switch (substate_id) {
+	case 1:
```
Do we expect some well named macros instead of raw 1, 2, ... values?

etienne-lms reviewed (commented):
Some nitpicking comments on coding style.

asm5878 commented on soc/st/stm32/stm32wbax/Kconfig:
```diff
@@ -18,3 +18,6 @@ config SOC_SERIES_STM32WBAX
 	select SOC_EARLY_INIT_HOOK
+imply PM_DEVICE if PM
+imply PM_DEVICE_RUNTIME if PM
+imply PM_DEVICE_SYSTEM_MANAGED if PM
```
Fixed.

asm5878 commented on soc/st/stm32/stm32wbax/hci_if/linklayer_plat_adapt.c:
```diff
@@ -10,7 +10,8 @@
 
-#include "scm.h"
+#include "app_conf.h"
+#include "bsp.h"
```
Done

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +22,80 @@
-	 * Disabling ICACHE is required before entering stop mode
-	 */
-	sys_cache_instr_disable();
+#define HSE_ON ((READ_BIT(RCC->CR, RCC_CR_HSEON) == RCC_CR_HSEON) ? 1UL : 0UL)
```
Done.

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +22,80 @@
-	LL_LPM_EnableDeepSleep();
+#if defined(CONFIG_PM_S2RAM)
+struct fpu_ctx_full fpu_state;
+SCB_Type backup_scb;
```
Done.

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +112,134 @@ static void set_mode_suspend_to_ram(void)
+		if (substate_id > 1) {
+			__HAL_FLASH_SET_LATENCY(flash_latency_backup);
+			while (__HAL_FLASH_GET_LATENCY() != flash_latency_backup) {
+			};
```
done

asm5878 committed:
dts: arm: st: wba: standby exit latency update

Updating standby exit latency time according worst case value
from stm32wba5 and stm32wba6 datasheets.

Signed-off-by: Alessandro Manganaro <alessandro.manganaro@st.com>

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +22,80 @@
 
 	while (LL_PWR_IsActiveFlag_ACTVOS() == 0) {
-	}
+	};
```
Done and verified.

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +23,51 @@
-	LL_PWR_ClearFlag_STOP();
-	LL_RCC_ClearResetFlags();
+#define STM32WBA_PM_STOP0 (1)
+#define STM32WBA_PM_STOP1 (2)
```
This macro definition was suggested by @etienne-lms. 
I haven't found any common definitions for the other stm32 families. 
So as suggested by @mathieuchopstm I have removed these macros.
Problem can be solved in an other PR in a systematic way for all stm32 mcus.

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +23,51 @@
-	/* Select mode entry : WFE or WFI and enter the CPU selected mode */
-	k_cpu_idle();
+	/* MCU exits from low power mode and set back the basepri */
+	__set_BASEPRI(basepri);
```
Applied and tested

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +84,134 @@ static void set_mode_suspend_to_ram(void)
+		set_mode_suspend_to_ram_enter();
+		/* Resume here */
+		set_mode_suspend_to_ram_exit();
+		__enable_irq();
```
Applied and tested.

mathieuchopstm commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +23,71 @@
 
-static void set_mode_suspend_to_ram(void)
+#if defined(CONFIG_PM_S2RAM)
+static void set_mode_suspend_to_ram_enter(void)
```
I would change this to `static int` and return the value of `arch_pm_s2ram_suspend()`...

asm5878 replied:
This optimization has not given the expected result. So I prefer to not go ahead with your suggestion.

mathieuchopstm commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +104,130 @@ static void set_mode_suspend_to_ram(void)
-		LOG_DBG("Suspend to RAM needs CONFIG_PM_S2RAM to be enabled");
-#endif
+	case PM_STATE_SUSPEND_TO_RAM:
+		set_mode_suspend_to_ram_enter();
+		/* Resume here */
+		set_mode_suspend_to_ram_exit();
```
...to check the return value here, and skip resume path if not needed:
```suggestion
		if (set_mode_suspend_to_ram_enter() >= 0) {
		   /* SoC entered in STANDBY mode successfully.
		    * Execution resumes here after wakeup.
		    */
		   set_mode_suspend_to_ram_exit();
		}
```
(this is more of an optimization than anything else)

asm5878 replied:
This optimization has not given the expected result. So I prefer to not go ahead with your suggestion.

msmttchr commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +104,130 @@ static void set_mode_suspend_to_ram(void)
 	/* Select standby mode */
 	LL_PWR_SetPowerMode(LL_PWR_MODE_STANDBY);
 
+	/* Save FPU state and configuration */
+	z_arm_save_fp_context(&fpu_state);
+	z_arm_save_scb_context(&scb_state);
```
See https://github.com/zephyrproject-rtos/zephyr/pull/93449

asm5878 replied:
Done.

mathieuchopstm commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +104,130 @@ static void set_mode_suspend_to_ram(void)
+	sys_cache_instr_disable();
+	/* If stop mode is greater than 1, manage flash latency and RAMCFG */
+	if (substate_id > 1) {
+		flash_latency_backup = __HAL_FLASH_GET_LATENCY();
+		if (flash_latency_backup < FLASH_LATENCY_1) {
+			/* Set flash latency to 1, since system will restart from HSI16 */
+			__HAL_FLASH_SET_LATENCY(FLASH_LATENCY_1);
+			while (__HAL_FLASH_GET_LATENCY() != FLASH_LATENCY_1) {
+			}
```
Nevermind. This only sets the latency if it was _under_ `LATENCY_1`, which is fine.

mathieuchopstm commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +23,71 @@
+	 * entry in STANDBY mode failed (return value of
+	 * this function is ignored for STOP mode entry).
+	 */
+	__set_BASEPRI(basepri);
+	__enable_irq();
+
```
We also need to revert the `LL_LPM_EnableDeepSleep()`:
```suggestion
	__set_BASEPRI(basepri);
	__enable_irq();
	
	/* Disable SLEEPDEEP at Cortex-M level */
        LL_LPM_EnableSleep();
```

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +104,130 @@ static void set_mode_suspend_to_ram(void)
-		scm_setup();
+	if (LL_PWR_IsActiveFlag_STOP() || !HSE_ON) {
+		/* Reconfigure the clock */
+		stm32_clock_control_init(NULL);
```
Done

asm5878 commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -18,69 +23,71 @@
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
+	__WFI();
```
Done.

mathieuchopstm reviewed (commented):
Otherwise LGTM 🚀

msmttchr reviewed (approved):
LGTM, thanks. For some reason I can’t “resolve conversations” (button is not shown for me). If you can, you can mark all my conversations resolved.

etienne-lms commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +108,127 @@ static void set_mode_suspend_to_ram(void)
+	case PM_STATE_SUSPEND_TO_RAM:
+		set_mode_suspend_to_ram_enter();
+		/* Resume here */
+		   set_mode_suspend_to_ram_exit();
```
nitpicking on indentation.

etienne-lms reviewed (commented):
Minor comments.

etienne-lms commented on soc/st/stm32/stm32wbax/power.c:
```diff
@@ -97,88 +84,134 @@ static void set_mode_suspend_to_ram(void)
 	default:
 		LOG_DBG("Unsupported power state %u", state);
-		break;
+		return;
```
ok

etienne-lms reviewed (commented):
LGTM AFAICT.

erwango reviewed (changes requested):
Otherwise LGTM

asm5878 commented on west.yml:
```diff
@@ -245,7 +245,7 @@ manifest:
         - hal
     - name: hal_stm32
-      revision: 611d54c237623ea71a367502bfbd736fc67ba15f
+      revision: pull/294/head
```
Conflict.... Rebase on going
